### DISCUSSION TITLE
PWM output not detected

### DISCUSSION THREAD

**Discussion by ericklil** [HomeSpan/HomeSpan#972] [General]
In the RGB LED example output for the PWM signal is set to be `DEV_RgbwLED(32,22,23,21);`
in [PwmPin.h](https://github.com/HomeSpan/HomeSpan/blob/master/src/src/extras/PwmPin.h) the Frequency is set at 5000.
using either a logic Analyser or scope no signal is found on the the above output pins.

Does the `ledcAttach(ledPin, freq, resolution);` function need to be called?

**Comment by HomeSpan** [comment]
No, you do not need to add other calls or functions.

I just tried the example 10 and it is working fine.  Why do you have an extra parameter in your function call to DEV_RgbwLED?  Also, what chip are you using, and are sure you connected everything to the correct pins?  Note the RGB LED must be common CATHODE where the common pin is grounded.

**Comment by ericklil** [comment]
Thank you, Indeed with a common Cathode RGB led connected it works, however I need to fee the signal to a led driver like the TI LM3414. to operate high powered common anode led's. Can the signal be inverted? 
Working on adding a white channel hence the W added in DEV_RgbwLED.

**Comment by HomeSpan** [comment]
If you just need an inverted digital output signal you can do that by adjusting the duty cycle to be *1-duty cycle*.  In other words, an inverted PWM with a duty cycle of 10% is the same as a regular PWM with a duty cycle of 90%. (well, the phase is different, but that should not matter when driving an LED).

However, to drive common-anode LED you probably need a PWM **input sink**, as opposed to an output.  I do not believe the ESP32 has an input PWM peripheral.

If your layout permits, you may want to consider using a NeoPixel LED instead of an RGB.  The NeoPixels are extremely bright and HomeSpan has an integrated library for controlling them.

**Comment by ericklil** [comment]
I have done extensive and successful work with Neopixels to the point of filing a Homekit compatible product with Apple, back in 2020. I am capable of sending an appropriate and usable PWM signal to Common anode-led driver with a simple code like:

```
// the number of the LED pin
const int ledPin = 32;  //  corresponds to GPIO32

// setting PWM properties
const int freq = 300;
const int resolution = 8;
 
void setup(){
  // configure LED PWM
  ledcAttach(ledPin, freq, resolution);

}
 
void loop(){

  // increase the LED brightness
  for(int dutyCycle = 0; dutyCycle <= 255; dutyCycle++){   
    ledcWrite(ledPin, dutyCycle);
    delay(15);
  }

  // decrease the LED brightness
  for(int dutyCycle = 255; dutyCycle >= 0; dutyCycle--){
    ledcWrite(ledPin, dutyCycle);   
    delay(15);
  }
}
```
I don't know to much about duty cycles. I was hoping to use the builtin capabilities of the ESP32 to achieve this. The LED fixtures I have in mind are common anode and the neopixels drivers including the PT 4115 chip don't appear to like a common anode.

**Comment by HomeSpan** [comment]
How are you connecting your common-anode RGB LED to the device?

**Comment by HomeSpan** [comment]
Also, one thing I just remembered is that I did add an invert flag to the HomeSpan PWM code a while back.  See the PWM page for details.  Maybe this will address your setup.

**Comment by ericklil** [comment]
> How are you connecting your common-anode RGB LED to the device?

Tested with a common anode RGB led. I am sending the same signal to a 350mA Constant current Driver board that I have. 

> Also, one thing I just remembered is that I did add an invert flag to the HomeSpan PWM code a while back. See the PWM page for details. Maybe this will address your setup.

With the invert flag checked in PwmPin.h, the logic is upside down, the signal is `high` when off, and `low` at full brightness  - in the HomeKit "dimmer"

Without the invert flag  I get no signal from the assigned pins. Should I attempt to create a variant of the PwmPin.h to address this, that would likely go way above my knowledge..

**Comment by HomeSpan** [comment]
What version of the Arduino-ESP32 Library are you using?

**Comment by ericklil** [comment]
> What version of the Arduino-ESP32 Library are you using?

3.1.0 ..as per your clear & kind instructions. I was flip flopping between my laptop and desktop, sharing the sketch and Homespun library on iCloud and somehow it would not update the sketch when I made changes in PwmPin.h. That is solved now. All is working as intended.

Would it be hard to add a White Channel to the RGB(W). Does HomeKit know how to handle that. I also hope to use less that have both a WW and CW led. 

Some makers have been able to create "lightshows" or "effects" although created in the EVE Home app they addressable as scenes in HomeKit. Does Homespan have support for that?

**Comment by HomeSpan** [comment]
Note the current version of HomeSpan has been tested only on Arduino-ESP32 version 3.0.X (as per the README page).  Version 3.1.0 of the Arduino-ESP32 was released by Espressif a few weeks ago.  I've just started testing and there two areas of incompatibility - one is directly related to PWM.  I've updated HomeSpan to conform to the new 3.1.0 library and will release in a few days.  In the interim, please only use Arduino-ESP32 3.0.7 (their latest 3.0.X branch).   The problem with the PWM in the new version is random at start-up.  PWM will either initialize normally, or won't initialize at all. This is because Espressif added a new parameter that lets you either initialize or de-initialize the LEDC driver from the same configuration structure.  They did not note this critical change in any of their migration documentation, which meant HomeSpan was not fully initializing the new LEDC configuration structure leaving the new flag that says "set to false to configure, or true to de-configure" to be a random value at start-up.  

Regarding WW and CW, HomeKit can operate a Light in "temperature mode" instead of RGB mode.   See the HomeSpan Pixel-RGBWC example for details.  The portion of the code that is Pixel-related is separate from the RGB and WC logic.

Regarding using EVE, yes, I've done exactly that --- there is nothing specific in EVE, but you can create custom Services and Characteristics in HomeSpan and EVE will support them (as opposed to the Apple Home App, which ignores them).

**Comment by ericklil** [comment]
> Note the current version of HomeSpan has been tested only on Arduino-ESP32 version 3.0.X (as per the README page). Version 3.1.0 of the Arduino-ESP32 was released by Espressif a few weeks ago. I've just started testing and there two areas of incompatibility - one is directly related to PWM.

Thank you that makes a lot of sense, of why I experienced erratic behavior. I will take a course reading now..

> Light in "temperature mode" instead of RGB mode.

Should start with expanding the pin assignment then. Curious how will the HSV algorithm know when to use the white led or not when mixing a color.. Having previously relied on NeoPixel library to figure that out, I would be grateful for a little guidance if this is even possible..

**Comment by HomeSpan** [comment]
The HSV algo I have is strictly a mapping to RGB.  In my example I use a single Pixel strip that support RGBWC.  I  use the RGB for one Accessory and then use the WC LEDs for a second Accessory operating in temperature space.  There is no interaction between the two.  

You could try to create your own algo to map RGB color space to RGBWC, but I don't know how that would really work in practice.

**Comment by HomeKidd** [comment]
> Some makers have been able to create "lightshows" or "effects" although created in the EVE Home app they addressable as scenes in HomeKit. Does Homespan have support for that?

That was me a few years ago. 😄 You don't need any special thing. As @HomeSpan  mentioned you can create custom  Characteristic in HomeSpan and  will show up in EVE 😄 But only applies to addressable LEDs

**Comment by HomeSpan** [comment]
I'm going to convert this to a discussion instead of an issue since the original problem has been solved.
